bolt-boost/src/server.rs

```diff
@@ -0,0 +1,582 @@
+    Json(delegation): Json<SignedDelegation>,
+) -> Result<impl IntoResponse, PbsClientError> {
+    info!(pubkey = %delegation.message.pubkey, validator_index = delegation.message.validator_index, "Delegating signing rights");
+    post_request(state, DELEGATE_PATH, &delegation).await?;
+    Ok(StatusCode::OK)
+}
+
+/// Revoke constraint submission rights from a BLS key.
+/// Spec: <https://chainbound.github.io/bolt-docs/api/builder#revoke>
+#[tracing::instrument(skip_all)]
+async fn revoke(
+    State(state): State<PbsState<BuilderState>>,
+    Json(revocation): Json<SignedRevocation>,
+) -> Result<impl IntoResponse, PbsClientError> {
+    info!(pubkey = %revocation.message.pubkey, validator_index = revocation.message.validator_index, "Revoking signing rights");
+    post_request(state, REVOKE_PATH, &revocation).await?;
+    Ok(StatusCode::OK)
+}
+
+/// Get a header with proofs for a given slot and parent hash.
+/// Spec: <https://chainbound.github.io/bolt-docs/api/builder#get_header_with_proofs>
+#[tracing::instrument(skip_all, fields(slot = params.slot))]
+async fn get_header_with_proofs(
+    State(state): State<PbsState<BuilderState>>,
+    Path(params): Path<GetHeaderParams>,
+    req_headers: HeaderMap,
+) -> Result<impl IntoResponse, PbsClientError> {
+    let ms_into_slot = ms_into_slot(params.slot, state.config.chain);
+    let max_timeout_ms = state.pbs_config().timeout_get_header_ms.min(
+        state
+            .pbs_config()
+            .late_in_slot_time_ms
+            .saturating_sub(ms_into_slot),
+    );
+
+    if max_timeout_ms == 0 {
+        warn!(
+            ms_into_slot,
+            threshold = state.pbs_config().late_in_slot_time_ms,
+            "late in slot, skipping relay requests"
+        );
+
+        return Ok(StatusCode::NO_CONTENT.into_response());
+    }
+
+    let (_, slot_uuid) = state.get_slot_and_uuid();
+
+    // prepare headers, except for start time which is set in `send_one_get_header`
+    let mut send_headers = HeaderMap::new();
+    // TODO: error handling
+    send_headers.insert(
+        HEADER_SLOT_UUID_KEY,
+        HeaderValue::from_str(&slot_uuid.to_string()).unwrap(),
+    );
+    send_headers.insert(
+        USER_AGENT,
+        get_user_agent_with_version(&req_headers).unwrap(),
+    );
+
+    let relays = state.relays();
+    let mut handles = Vec::with_capacity(relays.len());
+    for relay in relays.iter() {
+        handles.push(send_timed_get_header(
+            params,
+            relay.clone(),
+            state.config.chain,
+            state.pbs_config(),
+            send_headers.clone(),
+            ms_into_slot,
+            max_timeout_ms,
+        ));
+    }
+
+    let results = join_all(handles).await;
+    let mut relay_bids = Vec::with_capacity(relays.len());
+    let mut hash_to_proofs = HashMap::new();
+
+    // Get and remove the constraints for this slot
+    let constraints = state.data.constraints.remove(params.slot);
```

Comment:
not checking this Option will result in panicking if there are no constraints for this slot. is this safely handled?